File: js/src/providers/searchapi.ts
import type { HttpResponse } from "../http.js";
import type { SearchRequest, SearchResponse } from "../types.js";
import { finalize, result, type ProviderContext, type ProviderSpec } from "./base.js";

export const searchapi: ProviderSpec = {
  name: "searchapi",
  envKeys: ["SEARCHAPI_API_KEY", "SEARCHAPI_KEY"],
  defaultBaseUrl: "https://www.searchapi.io",
  capabilities: new Set(["country", "language", "news"]),

  prepare(ctx: ProviderContext, req: SearchRequest) {
    const params: Record<string, string | number | boolean | undefined> = {
      engine: req.searchType === "news" ? "google_news" : "google",
      q: req.query,
      num: req.maxResults,
    };
    if (req.country) params.gl = req.country.toLowerCase();
    if (req.language) params.hl = req.language;
    for (const [k, v] of Object.entries(req.extra)) params[k] = v as string;
    return {
      method: "GET",
      url: `${ctx.baseUrl}/api/v1/search`,
      headers: { Authorization: `Bearer ${ctx.apiKey ?? ""}`, Accept: "application/json" },
      params,
    };
  },

  parse(_ctx, res: HttpResponse, req: SearchRequest, elapsedMs: number): SearchResponse {
    const data = res.data ?? {};
    const items = data.organic_results ?? data.news_results ?? [];
    const results = items.map((item: any) =>
      result({
        title: item.title,
        url: item.link,
        snippet: item.snippet,
        source: item.source,
        publishedDate: item.date,
        score: item.position,
        raw: item,
      }),
    );
    const box = data.answer_box ?? {};
    const answer = box.answer ?? box.snippet;
    const total = data.search_information?.total_results;
    return finalize(searchapi, req, results, data, { answer, totalResults: total, elapsedMs });
  },
};

File: python/src/anysearch/providers/searchapi.py
"""SearchApi.io — real-time SERP API (Google and other engines)."""

from __future__ import annotations

import httpx

from .._http import PreparedRequest
from ..types import SearchRequest, SearchResponse
from .base import BaseProvider, Capability


class SearchApiProvider(BaseProvider):
    name = "searchapi"
    env_keys = ("SEARCHAPI_API_KEY", "SEARCHAPI_KEY")
    default_base_url = "https://www.searchapi.io"
    capabilities = frozenset(
        {
            Capability.COUNTRY,
            Capability.LANGUAGE,
            Capability.NEWS,
        }
    )

    def prepare(self, req: SearchRequest) -> PreparedRequest:
        params = {
            "engine": "google_news" if req.search_type == "news" else "google",
            "q": req.query,
            "num": req.max_results,
        }
        if req.country:
            params["gl"] = req.country.lower()
        if req.language:
            params["hl"] = req.language
        params.update(req.extra)
        return PreparedRequest(
            method="GET",
            url=f"{self.base_url}/api/v1/search",
            headers={"Authorization": f"Bearer {self.api_key}", "Accept": "application/json"},
            params=params,
        )

    def parse(self, response: httpx.Response, req: SearchRequest, elapsed_ms: float) -> SearchResponse:
        data = response.json()
        items = data.get("organic_results") or data.get("news_results") or []
        results = []
        for item in items:
            results.append(
                self._result(
                    title=item.get("title"),
                    url=item.get("link"),
                    snippet=item.get("snippet"),
                    source=item.get("source"),
                    published_date=item.get("date"),
                    score=item.get("position"),
                    raw=item,
                )
            )
        answer = None
        box = data.get("answer_box") or {}
        if isinstance(box, dict):
            answer = box.get("answer") or box.get("snippet")
        total = (data.get("search_information") or {}).get("total_results")
        return self._response(
            req, results, data, answer=answer, total_results=total, elapsed_ms=elapsed_ms
        )
